File: Leetcode543. Diameter of binary tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

using namespace std;
class Solution {
public:

    int res = 0;

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return res;
    }

    int dfs(TreeNode* node) {
        if(node==nullptr)
            return 0;

        // find the heights of the left and right subtree
        int left = dfs(node->left);
        int right = dfs(node->right);
        res = max(res, left+right);
        return 1+max(left, right);
    }
};

File: Leetcode543. Diameter of binary tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return res;
    }

    // returns height
    public int dfs(TreeNode node) {
        if(node==null)
            return 0;

        // get the heights of the left and right subtree
        int left = dfs(node.left);
        int right = dfs(node.right);
        res = Math.max(res, left+right);
        return 1 + Math.max(left, right);
    }
}
